File: cpp/Stacks/implement_a_queue_using_a_stack.cpp
#include <vector>
#include <optional>

class Queue {
public:
    Queue() {}

    void enqueue(int x) {
        enqueueStack.push_back(x);
    }

    void transferEnqueueToDequeue() {
        // If the dequeue stack is empty, push all elements from the enqueue stack
        // onto the dequeue stack. This ensures the top of the dequeue stack
        // contains the most recent value.
        if (dequeueStack.empty()) {
            while (!enqueueStack.empty()) {
                dequeueStack.push_back(enqueueStack.back());
                enqueueStack.pop_back();
            }
        }
    }

    std::optional<int> dequeue() {
        transferEnqueueToDequeue();
        // Pop and return the value at the top of the dequeue stack.
        if (!dequeueStack.empty()) {
            int val = dequeueStack.back();
            dequeueStack.pop_back();
            return val;
        }
        return std::nullopt;
    }

    std::optional<int> peek() {
        transferEnqueueToDequeue();
        // Return the value at the top of the dequeue stack without removing it.
        if (!dequeueStack.empty()) {
            return dequeueStack.back();
        }
        return std::nullopt; 
    }

private:
    std::vector<int> enqueueStack;
    std::vector<int> dequeueStack;

};

File: csharp/Stacks/ImplementAQueueUsingAStack.cs
public class Queue
{
    private Stack<int> _enqueueStack;
    private Stack<int> _dequeueStack;

    public Queue()
    {
        _enqueueStack = new Stack<int>();
        _dequeueStack = new Stack<int>();
    }

    public void Enqueue(int x)
    {
        _enqueueStack.Push(x);
    }

    private void TransferEnqueueToDequeue()
    {
        // If the dequeue stack is empty, push all elements from the enqueue stack
        // onto the dequeue stack. This ensures the top of the dequeue stack
        // contains the most recent value.
        if (_dequeueStack.Count == 0)
        {
            while (_enqueueStack.Count > 0)
                _dequeueStack.Push(_enqueueStack.Pop());
        }
    }

    public int? Dequeue()
    {
        TransferEnqueueToDequeue();

        // Pop and return the value at the top of the dequeue stack.
        return _dequeueStack.Count > 0 ? _dequeueStack.Pop() : null;
    }

    public int? Peek()
    {
        TransferEnqueueToDequeue();
        return _dequeueStack.Count > 0 ? _dequeueStack.Peek() : null;
    }
}
